### PR TITLE
Bulk Allocations - HCAP-1282/1283/1297

### PR DESCRIPTION
**Pull request by mmumby**
# Work completed
- Ability to add allocations in bulk
- Added multi-select to the site View table if MOH and 'Set Allocation' button if MOH
- Added bulk allocation form
- added bulk allocation POST API
- Added query params to GetAllPhases to conditionally include and array of allocations
- Added bulkAllcation service that checks for the existence of an allocation and either updates or creates a new one.
- Added logic for UI `acknowledgement` when overriding existing allocations 

## Follow-up work
- Test suites (e2e, server, cypress)

### UI
Screen Shot 2023-03-03 at 1 08 20 PM
Screen Shot 2023-03-03 at 1 08 33 PM
- The alert has a scroll if over 100px
Screen Shot 2023-03-03 at 1 08 59 PM

### Video of feature/errors/validations

https://user-images.githubusercontent.com/25125247/222812362-1f7ecddb-47af-465d-b9cc-2f5fa7fc6e5f.mov

### CODE REVIEW

File: client/src/components/modal-forms/BulkAllocationForm.js
Comment (fw-noel) [inline client/src/components/modal-forms/BulkAllocationForm.js:52]: This looks like it could be potentially optimized further. Lots of array functions going on that makes it a little harder to read and I wonder if it could be shortened. Don't know how yet, but I'm wondering if maybe some sort of dictionary of allocations that we could pull from.

Feel free to disagree though!

File: client/src/components/modal-forms/BulkAllocationForm.js
Comment (fw-noel) [inline client/src/components/modal-forms/BulkAllocationForm.js:62]: Not sure I understand why this needs to be in state AND in the form values? Is it in values because it's being submitted as a payload on submit?

File: client/src/pages/private/SetBulkAllocation.js
Comment (fw-noel) [inline client/src/pages/private/SetBulkAllocation.js:31]: Github Actions has a format check here

File: client/src/services/phases.js
Comment (fw-noel) [inline client/src/services/phases.js:30]: Necessary? Can't we just remove the trailing forward slash and add ${queryString} at the end? Or would that go `api/v1/phasenull`?

File: server/services/phase.ts
Comment (fw-noel) [inline server/services/phase.ts:123]: Neat

File: server/validators/allocation.ts
Comment (fw-noel) [inline server/validators/allocation.ts:51]: I think we can keep these (existingAllocations and acknowledgement) as a strictly frontend validation, I don't think it makes a lot of sense in a post request

File: server/services/allocations.ts
Comment (fw-noel) [inline server/services/allocations.ts:65]: This feels like a lot of requests as well. Can we write up a service for getting allocation for multiple sites in a phase? Or maybe grabbing all and filtering based on siteId?

File: server/routes/phase.js
Comment (fw-noel) [inline server/routes/phase.js:2]: Github Actions says this isn't used?

**Review by fw-noel — COMMENTED** [review]
Looks alright! I had a few nitpicks and a few optimization pieces.

File: client/src/services/phases.js
Comment (mmumby) [reply client/src/services/phases.js:30]: `api/v1/phasenull` is exactly what it does! I was also hoping to simplify here

File: client/src/components/modal-forms/BulkAllocationForm.js
Comment (mmumby) [reply client/src/components/modal-forms/BulkAllocationForm.js:62]: The `valuesexistingAllocations` is just a boolean, There didn't seem to be a need to have the entire array in the payload for the yup validation. the state variable `existingAllocations` is being used to display the site/allocations in the alert message.

File: client/src/services/phases.js
Comment (mmumby) [reply client/src/services/phases.js:30]: alternatively, I could default it to an empty string `queryString = ' ' ` and remove the const url

File: server/routes/phase.js
Comment (mmumby) [reply server/routes/phase.js]: removed! not sure why my linting didn't pick this up on the commit hook

File: client/src/pages/private/SetBulkAllocation.js
Comment (mmumby) [reply client/src/pages/private/SetBulkAllocation.js]: removed!

File: server/validators/allocation.ts
Comment (mmumby) [reply server/validators/allocation.ts]: good call, removed

File: client/src/services/phases.js
Comment (fw-noel) [reply client/src/services/phases.js:30]: Nah, this is fine!

File: client/src/components/modal-forms/BulkAllocationForm.js
Comment (mmumby) [reply client/src/components/modal-forms/BulkAllocationForm.js]: It can be shortened. It does require two array.filters/finds. I have split it into two consts to make it easier to read: 
```
 // get all allocations associated to the selected phase
    const phaseAllocations = phases.find(({ id }) => id === phaseId)?.allocations;
    // filter the allocations and return all that are associated to the sites selected.
    const phaseSiteAllocations = phaseAllocations.filter(({ site_id }) =>
      siteIds.includes(site_id)
    );
```

I was then mapping to format the object as I needed the site name to be included in the object, (Initially, I was passing this array to the yup validator/BE so I intentionally made the object smaller)  But that isn't necessary anymore, and I can read the siteName a different way. (created an object hash)

File: server/services/allocations.ts
Comment (mmumby) [reply server/services/allocations.ts]: good call - I have updated it to make 1 query to `getAllocationsForSites`, I then check to ensure the site being referenced exists in the array or existing allocations `allocationFound`, if it is, I update allocation, if not I create allocation. Let me know if this solution is more aligned with what you were thinking.

**Review by fw-noel — APPROVED** [review]
Approved! The only other thing is that it might be worth combining the POST set allocation and POST bulk allocation endpoints now, so we won't have to worry about it later

File: client/src/utils/date.js
Comment (rhattie-ey) [inline client/src/utils/date.js:13]: Not a fan of these names. All strings are text strings... the difference is the formatting. The names of these functions should reflect that.

File: server/routes/allocation.js
Comment (rhattie-ey) [inline server/routes/allocation.js:90]: At some point we might wanna roll this into the main allocation route, kind of like what we discussed with the other upcoming bulk work. This is fine for now though.

File: server/services/phase.ts
Comment (rhattie-ey) [inline server/services/phase.ts:114]: Document what this new query parameter does with JSDoc, and add a type. Neither is immediately clear at the moment, and ideally we should be striving for as much strictly defined typing as possible.

The name is also a potential source of confusion - I'd expect a parameter called `query` to be something like a SQL query in this region of the codebase, but here it looks like, presumably, query parameters from a request? I haven't had a chance to check in-IDE or look in greater depth, but that's what it looks like from here.

Considering all you're doing with this property is reading a bool from it, I'd highly suggest just replacing `query` with an `includeAllocations` parameter, and adjusting your call to `getAllPhases` to pass it that specific property of the query object.

**Review by rhattie-ey — COMMENTED** [review]
Looks good, although there is some potential room for cleanup.

File: server/routes/allocation.js
Comment (mmumby) [reply server/routes/allocation.js:90]: That was mentioned in the dev channel - I did move it OOS at this point due to the logic of post vs patch whereas the example we discussed only involved post. If we merge this feature into the existing POST, some updates/decisions will need to be made regarding the existing UI. (ie if for some reason they try to post a new allocation and one exists, do we allow/override and inform the user, or prevent that from occurring as it currently is implemented). 
If there is time after testing is completed I can look into merging the two.

File: client/src/utils/date.js
Comment (mmumby) [reply client/src/utils/date.js]: Updated to differentiate between long vs short form. If you have a better name suggestion let me know!

File: server/services/phase.ts
Comment (mmumby) [reply server/services/phase.ts]: Also - I created a `phase` type as it was a different response than `sitePhase`, which exposed the inconsistencies of the response in regards to casing (snake vs camel). I'm making it OOS for this ticket as the API was already implemented and there is still lots of testing to complete before EOS, but it's a Good Candidate for a tech debt ticket to update the casing, and decide on preferred formatting for APIs moving forward. (again, if there is time after testing I can update)

File: client/src/utils/date.js
Comment (rhattie-ey) [reply client/src/utils/date.js]: Looks good enough to me!

File: server/services/phase.ts
Comment (rhattie-ey) [inline server/services/phase.ts:132]: Typo in description - with formatted what, exactly?

File: server/services/phase.ts
Comment (rhattie-ey) [reply server/services/phase.ts]: Good point in regards to casing inconsistency, let's have a discussion about how we want to handle that later, and make a ticket for it. I like the new changes you made here.

File: server/services/phase.ts
Comment (mmumby) [reply server/services/phase.ts]: fixed! the sun was shining bright on my screen this morning, tough to see the comments.

**Review by fw-noel — APPROVED** [review]

**Review by rhattie-ey — APPROVED** [review]

**Pull request merged**
